### ISSUE TITLE
Implement DeviceRegistration Component

### ISSUE THREAD

**Issue by aghalarp** [openpowerquality/opq#3]
Please do work for this task in a branch called issue-3.

Implement a device registration system that fulfills the following use case:

1. The user, having received a new OPQBox, goes to the device registration page where they must enter their device's serial number.

2. If the serial number is already associated with an account, then an error message is displayed.

3. If the serial number is not yet associated with an account, and it is a valid serial number, then the system requests an email address from the visitor.

4. Once an email address is supplied, the system generates an email to that address with a link that enables the user to register themselves as the owner of that device. Clicking the link enables them to specify a password.

5. They can then login with their email address and password.

**Comment by philipmjohnson** [comment]
Single person "ownership" is complex to implement, because (a) you need to define an additional "viewer" role so that more than one person can see the same box, and (b) you need to define a way to change ownership. 

I suggest that you replace the idea of single person "owners" with multi-person "viewers"; anyone with the serial number can become a viewer.

Along with this, instead of serial numbers being easily guessable small integers (5), we might want to start making them a somewhat longer string ("356-AB-23").

**Comment by anthonyjchriste** [comment]
We currently publicly advertise the device id. I'm concerned that if all it takes is a serial number to subscribe  to the box, then it will be impossible to have secure boxes (assuming there is a use case for wanting to hide certain boxes). 

We may want to differentiate between "box id" and "box serial number" if the box id is going to be displayed publicly.

**Comment by philipmjohnson** [comment]
We don't have to publicly advertise the device ID with the following changes:

*  The Box entity has a "name", "location" and "description" field, that are user-configurable. The name field must be unique. 
* When a box initially connects, OPQView assigns a default name based upon the timestamp of the first measurement, ie "Box-2017-12-16T16:32:32Z-10:00". This makes it possible for a new user to identify their box in the public listing of boxes. 
* When the user connects to OPQView, they can enter the serial number to manage the name, location, and description fields. In particular, they could rename the box from the default name to something more description (yet still unique), i.e. "PJ-Kailua".

With this approach, we don't have to differentiate between box_id and box serial number.

**Comment by aghalarp** [comment]
I've updated the [data model](https://open-power-quality.gitbooks.io/open-power-quality-manual/content/datamodel/description.html#opq_boxes) to reflect these ideas, mainly:

- box_id is now a string (previously integer)
- Added a "name", "description", and "location" field (an object with a "zipcode" sub-field) to the Box entity

Just to clarify (for myself) this entire process of hooking up a new box:

- A user receives a new box, which has an internal box_id that we've assigned to it. This box_id is no longer a simple integer but a more complex string.
- The user hooks the box up to their router, and Makai/Mauka starts receiving data from it (which includes its box_id)
- Makai/Mauka queries the "opq_boxes" collection and does not find an existing box with that box_id - therefore it creates a new box document and assigns it the default name as described above.
- The user logs on to OPQView, sees a public listing of boxes (by their "name" field), selects the box they are interested in and enters its box_id as if it were a kind of password. If the "password" is correct, we can then mark that user as having permission to access/view that box (see the "boxes" array in the "users" collection).
- The user is then redirected to some kind of device administration page in OPQView, where they can change its name, add a description, etc.
